### PR TITLE
fix: adapt kubernetes event

### PR DESCRIPTION
**Pull request by xjh1996**
## What does this PR do?
adapt kubernetes event which lastTimestamp is empty or count is empty
```yaml
apiVersion: v1
items:
- action: Scheduling
  apiVersion: v1
  eventTime: "2021-12-22T11:39:51.388829Z"
  firstTimestamp: null
  involvedObject:
    apiVersion: v1
    kind: Pod
    name: insufficent-697c868bd6-xvjxp
    namespace: sjh-ns
    resourceVersion: "390502"
    uid: 11f3fe9f-f918-482d-9302-09f7dc69f450
  kind: Event
  lastTimestamp: null
  message: '0/3 nodes are available: 3 Insufficient cpu.'
  metadata:
    creationTimestamp: "2021-12-22T11:39:51Z"
    name: insufficent-697c868bd6-xvjxp.16c310d3dada4017
    namespace: sjh-ns
    resourceVersion: "5095"
    uid: 956af82e-fb37-44a3-b23e-ab3e158d2e90
  reason: FailedScheduling
  reportingComponent: default-scheduler
  reportingInstance: default-scheduler-mc70ksqfqtofq1s4t4njg
  source: {}
  type: Warning
- action: Scheduling
  apiVersion: v1
  eventTime: "2021-12-22T11:39:51.434179Z"
  firstTimestamp: null
  involvedObject:
    apiVersion: v1
    kind: Pod
    name: insufficent-697c868bd6-xvjxp
    namespace: sjh-ns
    resourceVersion: "390503"
    uid: 11f3fe9f-f918-482d-9302-09f7dc69f450
  kind: Event
  lastTimestamp: null
  message: '0/3 nodes are available: 3 Insufficient cpu.'
  metadata:
    creationTimestamp: "2021-12-22T11:39:51Z"
    name: insufficent-697c868bd6-xvjxp.16c310d3dd8dc308
    namespace: sjh-ns
    resourceVersion: "5202"
    uid: 796cd76b-fc90-427f-a899-2587c181c180
  reason: FailedScheduling
  reportingComponent: default-scheduler
  reportingInstance: default-scheduler-mc70ksqfqtofq1s4t4njg
  series:
    count: 18
    lastObservedTime: "2021-12-22T12:06:01.048247Z"
  source: {}
  type: Warning
kind: List
metadata:
  resourceVersion: ""
  selfLink: ""
```

## Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Author's Checklist

- [ ]

## Related issues

-

### CODE REVIEW

**Review by mtojek — COMMENTED** [review]
Hi @xjh1996,  you opened this PR against the old branch. Would you mind opening it against the latest main branch? Does it make sense?

**Comment by xjh1996** [comment]
Hello @mtojek，I need to use this branch to avoid introducing other modifications in our environment.

**Comment by mtojek** [comment]
I'm afraid that you have to use your own fork in this case. We don't actively backport/support old branches. If you want to add a feature or code modification, feel free to start with the main branch and we can backport it to other places.

**Comment by xjh1996** [comment]
> I'm afraid that you have to use your own fork in this case. We don't actively backport/support old branches. If you want to add a feature or code modification, feel free to start with the main branch and we can backport it to other places.

@mtojek ok，I will rebase my code.

File: metricbeat/module/kubernetes/event/event.go [hunk 1]
```diff
@@ -193,8 +205,14 @@ func generateMapStrFromEvent(eve *kubernetes.Event, dedotConfig dedotConfig) com
 
 	tsMap := make(common.MapStr)
 
-	tsMap["first_occurrence"] = kubernetes.Time(&eve.FirstTimestamp).UTC()
-	tsMap["last_occurrence"] = kubernetes.Time(&eve.LastTimestamp).UTC()
+	lastTimestamp := kubernetes.EventLastTimestamp(eve)
+	tsMap["first_occurrence"] = func() time.Time {
+		if eve.FirstTimestamp.IsZero() {
+			return kubernetes.Time(&lastTimestamp).UTC()
```
Comment (ChrsMark) [inline metricbeat/module/kubernetes/event/event.go:211]: I'm not sure if this is correct. If an event has no `FirstTimestamp` attached then we fallback to `lastTimestamp` so this means that each time the event occurs we emit data with different `first_occurrence` which is always equal to the current `lastTimestamp`. Is this something that is intentional, and if so why?

Maybe it's better here to use `EventTime` as a fallback?

File: libbeat/common/kubernetes/types.go [hunk 2]
```diff
@@ -116,3 +116,17 @@ func ContainerIDWithRuntime(s PodContainerStatus) (string, string) {
...
+
+func EventLastTimestamp(eve *Event) metav1.Time {
+	lastTimestamp := eve.LastTimestamp
+	if lastTimestamp.IsZero() {
+		if eve.Series != nil && !eve.Series.LastObservedTime.IsZero() {
+			lastTimestamp = metav1.Time(eve.Series.LastObservedTime)
+		} else if !eve.EventTime.IsZero() {
+			lastTimestamp = metav1.Time(eve.EventTime)
+		} else {
+			lastTimestamp = eve.CreationTimestamp
```
Comment (ChrsMark) [inline libbeat/common/kubernetes/types.go:128]: Does this field actually exist? I cannot find on the version of the client library we use (v1.21): https://github.com/kubernetes/api/blob/release-1.21/core/v1/types.go#L5463

File: libbeat/common/kubernetes/types.go [hunk 3]
```diff
@@ -116,3 +116,17 @@ func ContainerIDWithRuntime(s PodContainerStatus) (string, string) {
...
 	return "", ""
 }
+
+func EventLastTimestamp(eve *Event) metav1.Time {
+	lastTimestamp := eve.LastTimestamp
+	if lastTimestamp.IsZero() {
+		if eve.Series != nil && !eve.Series.LastObservedTime.IsZero() {
+			lastTimestamp = metav1.Time(eve.Series.LastObservedTime)
+		} else if !eve.EventTime.IsZero() {
+			lastTimestamp = metav1.Time(eve.EventTime)
```
Comment (ChrsMark) [inline libbeat/common/kubernetes/types.go:126]: Here we fallback to a value that is described as `Time when this Event was first observed`, is this something that is actually correct? `EventTime` is mostly a good fallback if `FirstTimestamp` is not present.

**Review by ChrsMark — COMMENTED** [review]
Hey @xjh1996, thank you very much for your contribution!

I have some concerns about the implementation but the direction is in general sth that we can adopt so let me provide some feedback and we can discuss on it and iterate on the PR:

1. +1 on using `eve.Series.Count` as fallback when `eve.Count` is not present.
2. I would suggest using `eve.EventTime` as fallback when `FirstTimestamp` is not present
3. +1 on using `eve.Series.LastObservedTime` as fallback when  `eve.LastTimestamp` is not present.

Let me know what you think :).

**Comment by ChrsMark** [comment]
@xjh1996 I'm going to close this PR for now and keep the discussion going at https://github.com/elastic/beats/issues/31126. Let us know there your opinions and if you still want to contribute to this you can open the PR after we decide on the changes approach at https://github.com/elastic/beats/issues/31126.

**Pull request closed without merging**